build-tools/automation/azure-pipelines.yml

```diff
@@ -0,0 +1,215 @@
+      condition: always()
+
+    - task: PublishPipelineArtifact@0
+      displayName: upload artifacts
+      inputs:
+        artifactName: prepare-build-results
+        targetPath: $(Build.ArtifactStagingDirectory)/results
+      condition: always()
+
+- stage: mac_build
+  displayName: Mac
+  dependsOn: prepare
+  jobs:
+  - job: mac_build
+    displayName: Build
+    pool: Hosted Mac Internal
+    timeoutInMinutes: 240
+    cancelTimeoutInMinutes: 5
+    variables:
+      Configuration: Release
+      JavaSdkDirectory: /Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/
+      JAVA_HOME: $(JavaSdkDirectory)
+    steps:
+    - task: DownloadPipelineArtifact@1
+      inputs:
+        artifactName: $(BundleArtifactName)
+        downloadPath: $(System.DefaultWorkingDirectory)
+
+    - script: make prepare-external-git-dependencies CONFIGURATION=$(Configuration)
+      displayName: make prepare-commercial
+      condition: eq(variables['XA.Commercial.Build'], 'true')
+      env:
+        GH_AUTH_SECRET: $(Github.Token)
+
+    - task: provisionator@2
+      condition: and(succeeded(), eq(variables['XA.Commercial.Build'], 'true'))
+      inputs:
+        provisionator_uri: $(provisionator-uri)
+        github_token: $(GitHub.Token)
+        provisioning_script: $(System.DefaultWorkingDirectory)/external/monodroid/build-tools/provisionator/profile.csx
+        provisioning_extra_args: -vv
+
+    - script: make prepare-deps V=1 CONFIGURATION=$(Configuration) MSBUILD_ARGS="$(AutoProvisionArgs)"
+      displayName: make prepare-deps
+
+    - script: make prepare jenkins V=1 CONFIGURATION=$(Configuration) MSBUILD_ARGS="$(AutoProvisionArgs) /p:BundleRootPath=$(System.DefaultWorkingDirectory)"
+      displayName: make prepare jenkins
+
+    - script: make create-installers V=1 CONFIGURATION=$(Configuration)
+      displayName: make create-installers
+
+    - script: |
+        mkdir -p $(System.DefaultWorkingDirectory)/unsigned-installers
+        cp bin/Build$(Configuration)/*.vsix $(System.DefaultWorkingDirectory)/unsigned-installers
+        cp bin/Build$(Configuration)/*.pkg $(System.DefaultWorkingDirectory)/unsigned-installers
+      displayName: copy build artifacts
+
+    - task: PublishPipelineArtifact@0
+      displayName: upload unsigned installers
+      inputs:
+        artifactName: unsigned
+        targetPath: $(System.DefaultWorkingDirectory)/unsigned-installers
+
+    - task: MSBuild@1
+      displayName: package build results
+      inputs:
+        solution: build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks.csproj
+        configuration: $(Configuration)
+        msbuildArguments: /t:ZipBuildStatus /p:ResultPackageOutputPath=$(Build.ArtifactStagingDirectory)
+      condition: always()
+
+    - task: PublishPipelineArtifact@0
+      displayName: upload artifacts
+      inputs:
+        artifactName: mac-build-results
+        targetPath: $(Build.ArtifactStagingDirectory)
```

Comment:
This kinda confuses me.  On line 63, the `targetPath` is `(Build.ArtifactStagingDirectory)/results`; note `/results`.  Here, it's just `(Build.ArtifactStagingDirectory)`; note the *lack* of `/results`.

Should these be the same?  Should they differ?  Do they need to differentiations because, at present, the filename is always the same?

(Un?)related: where are these files uploaded *to*?  What's the final URL look like?  Can files be (accidentally?  easily?) overwritten as part of these publishing commands?  What happens with *all* of `create_bundle`, `mac_build`, and (eventual) `windows_build` need to upload some form of `xa-build-status.zip`?

Reply:
The reason the extra `results` folder is used on line 63 is because the `$(Build.ArtifactStagingDirectory)` directory was used previously to publish the `bundle` artifact. That folder still contains a `bundle.zip` file at this stage, which would be uploaded again as a separate artifact if this folder was used again.

The `artifactName` input is what is providing "uniqueness" to the build/test result artifacts. Here's a screenshot of what our artifacts currently look like for this pipeline:

![image](https://user-images.githubusercontent.com/2000163/58337278-7392f500-7e13-11e9-89b5-9f9b7ecc7c05.png)
![image](https://user-images.githubusercontent.com/2000163/58337290-7aba0300-7e13-11e9-9674-135ca37bff92.png)

All of these artifacts have unique URIs which are based on the build instance/commit that they are associated with. To your previous point however, I can re-introduce the detailed suffix for build/result file names to make it easier to deal with some of these files on a local file system.